File: JavaScript/DSA/Queue.js
// A Queue is a linear data structure that follows a First-In-First-Out (FIFO) approach

class Queue {
  constructor() {
    this.items = [];
  }

  // add an item to the end of the queue
  enqueue(item) {
    this.items.push(item);
  }

  // remove an item from the front of the queue
  dequeue() {
    if (this.isEmpty()) {
      return "Underflow";
    }
    return this.items.shift();
  }

  // get the front item of the queue without removing it
  peek() {
    if (this.isEmpty()) {
      return "No elements in Queue";
    }
    return this.items[0];
  }

  // check if the queue is empty
  isEmpty() {
    return this.items.length == 0;
  }

  // get the size of the queue
  size() {
    return this.items.length;
  }

  // print the elements of the queue
  print() {
    let str = "";
    for (let i = 0; i < this.items.length; i++) {
      str += this.items[i] + " ";
    }
    return str;
  }
}

// usage

const queue = new Queue();

console.log(queue.isEmpty()); // Output: true

queue.enqueue(10);
queue.enqueue(20);
queue.enqueue(30);

console.log(queue.print()); // Output: 10 20 30

console.log(queue.size()); // Output: 3

console.log(queue.peek()); // Output: 10

console.log(queue.dequeue()); // Output: 10

console.log(queue.print()); // Output: 20 30

File: C&C++/DSA/Queue.cpp
/*
Queue is a linear data structure in C++ that stores a collection of elements in a sequence, following the FIFO (First-In-First-Out) principle.
*/

#include <iostream>
#include <queue>

using namespace std;

// Create (Enqueue) function
void enqueue(queue<int> &q, int val)
{
    q.push(val);
    cout << "Element " << val << " added to the queue." << endl;
}

// Read (Peek) function
void peek(queue<int> q)
{
    if (!q.empty())
    {
        cout << "The first element in the queue is: " << q.front() << endl;
    }
    else
    {
        cout << "The queue is empty." << endl;
    }
}

// Update function
void update(queue<int> &q, int old_val, int new_val)
{
    if (!q.empty())
    {
        queue<int> temp;
        bool found = false;

        while (!q.empty())
        {
            int val = q.front();
            q.pop();

            if (val == old_val)
            {
                found = true;
                temp.push(new_val);
            }
            else
            {
                temp.push(val);
            }
        }

        if (!found)
        {
            cout << "Element " << old_val << " not found in the queue." << endl;
        }
        else
        {
            q = temp;
            cout << "Element " << old_val << " updated to " << new_val << "." << endl;
        }
    }
    else
    {
        cout << "The queue is empty." << endl;
    }
}

// Delete (Dequeue) function
void dequeue(queue<int> &q)
{
    if (!q.empty())
    {
        cout << "Element " << q.front() << " removed from the queue." << endl;
        q.pop();
    }
    else
    {
        cout << "The queue is empty." << endl;
    }
}

int main()
{
    queue<int> q;

    // Create (Enqueue) operation
    enqueue(q, 10);
    enqueue(q, 20);
    enqueue(q, 30);

    // Read (Peek) operation
    peek(q);

    // Update operation
    update(q, 20, 25);

    // Read (Peek) operation after update
    peek(q);

    // Delete (Dequeue) operation
    dequeue(q);

    // Read (Peek) operation after delete
    peek(q);

    return 0;
}
